### PR TITLE
Pre-render properties that have format specifiers

### LINKED ISSUE #2: Pre-render tokens with format specifiers
https://github.com/datalust/nlog-targets-seq/blob/dev/src/NLog.Targets.Seq/LogEventInfoFormatter.cs#L72

So that the message template can be accurately rendered on the remote machine, any formatted tokens in the message template need to be pre-rendered.

E.g.:

```csharp
logger.Info("The number is {N:000.00}", 42);
```

Here we want to retrieve the format `000.00` associated with the `N` property.

@304NotModified any clue how this can be achieved?

### PR DESCRIPTION
**Pull request by nblumhardt**
Fixes #2

### CODE REVIEW

File: src/NLog.Targets.Seq/LogEventInfoFormatter.cs [hunk 1]
```diff
@@ -49,58 +51,87 @@ static LogEventInfoFormatter()
...
+        public static void ToCompactJson(LogEventInfo logEvent, TextWriter output, IList<SeqPropertyItem> properties)
         {
-            if (loggingEvent == null) throw new ArgumentNullException(nameof(loggingEvent));
+            if (logEvent == null) throw new ArgumentNullException(nameof(logEvent));
             if (output == null) throw new ArgumentNullException(nameof(output));
             if (properties == null) throw new ArgumentNullException(nameof(properties));
 
+            // Without this call, logEventInfo.Properties behaves erratically when templates are used; stll trying
+            // to track down the cause (needs some debugging into NLog).
+            logEvent.FormattedMessage?.ToString();
```
Comment (nblumhardt) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: @304NotModified have I hit a bug here? Without this line (which I assume triggers some internal message template handling), later calls to `Properties.ContainsKey("X")` return false even with `X` in the template.

File: src/NLog.Targets.Seq/LogEventInfoFormatter.cs [hunk 2]
```diff
@@ -49,58 +51,87 @@ static LogEventInfoFormatter()
...
+            output.Write(logEvent.TimeStamp.ToUniversalTime().ToString("O"));
+
+            string message = null;
+            Template template = null;
+            try
+            {
+                if (logEvent.Message != null)
+                {
+                    message = logEvent.Message;
+                    template = TemplateParser.Parse(logEvent.Message);
```
Comment (nblumhardt) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: It'd be nice to avoid this and piggy-back off of work NLog is already doing.

**Comment by nblumhardt** [comment]
Merging to get the published version up to a reasonable level of functionality, but keen to follow up on whatever's causing the failure noted above.

File: src/NLog.Targets.Seq/LogEventInfoFormatter.cs (hunk 1 above)
Comment (nblumhardt) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: The problem appears to be that `LogEvent.CalcFormattedMessage()` (private) actually populates the `Properties` collection:

https://github.com/NLog/NLog/commit/67368ec3f0840cb7c1c312a51225c50ff7abb84a#diff-fe47bd426efb3db67d05e56b297587c6R509

Would it be less of a footgun if the first call to read `Properties` invoked this logic? Formatting wouldn't be required, but unless the template is fully processed, `Properties` won't be complete.

Comment (304NotModified) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: need some more context here ;) is this NLog 4.5?

Comment (304NotModified) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: is this still an issue as this PR is merged?

Comment (nblumhardt) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: Yes, this is 4.5 :-) Brand new target for it!

The problem is that it's possible to access `LogEventInfo.Properties` without ever having triggered the internal calc method, thus the collection will be empty when instead it should have been populated from the template and params.

Shall I raise an issue on NLog for this one too? Cheers!

Comment (304NotModified) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: > Shall I raise an issue on NLog for this one too? Cheers!

yes please, easier to track it. Thanks!

File: src/NLog.Targets.Seq/LogEventInfoFormatter.cs (hunk 2 above)
Comment (304NotModified) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: We should expose `Template`, isn't it?

Comment (nblumhardt) [inline src/NLog.Targets.Seq/LogEventInfoFormatter.cs]: Yes, I think that's the way to go.

**Pull request merged**